### PR TITLE
patch: Add the beginnings of a patch utility

### PR DESCRIPTION
**Pull request by shannonbooth**
This is still very bare bones, and there is _much_ more to still
handle. However, this implements enough functionality to parse a single
unified patch read from stdin, and apply it to a file.

image

image

image

### CODE REVIEW

File: Userland/Libraries/LibDiff/Hunks.cpp [hunk 1]
```diff
@@ -10,45 +10,100 @@
...
+    return location;
+}
 
-    while (line_index < diff_lines.size()) {
-        if (diff_lines[line_index][0] == '@') {
-            current_location = parse_hunk_location(diff_lines[line_index]);
-            ++line_index;
+bool Parser::consume_line_number(size_t& number)
+{
+    auto line = consume_until([](char ch) { return ch < '0' || ch > '9'; });
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibDiff/Hunks.cpp:50]: This can be written more simply:
```suggestion
    auto line = consume_while(is_ascii_digit);
```

(You'll need to include `AK/CharacterTypes.h` if it's not already included.)

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
One nit-pick.

File: Userland/Libraries/LibDiff/Hunks.cpp (hunk 1 above)
Comment (shannonbooth) [reply Userland/Libraries/LibDiff/Hunks.cpp]: Thanks @AtkinsSJ , that's much nicer!

**Comment by shannonbooth** [comment]
pushed rebase to fix conflicts

**Review by AtkinsSJ — APPROVED** [review]
Looks good! Might be a good idea to add some tests for `patch` if we want to start relying on it in places, but that can be a future PR.

**Comment by shannonbooth** [comment]
@AtkinsSJ thanks! And yup definitely, there are a suprising number of edge cases to handle
